File: android/lib/src/main/java/me/anasmusa/learncast/lib/screen/player/PlayerActionSheet.kt
package me.anasmusa.learncast.lib.screen.player

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import me.anasmusa.learncast.Strings
import me.anasmusa.learncast.data.model.DownloadState
import me.anasmusa.learncast.lib.AppTheme
import me.anasmusa.learncast.lib.component.SheetMenuButton
import me.anasmusa.learncast.lib.theme.icon.Delete
import me.anasmusa.learncast.lib.theme.icon.DoneAll
import me.anasmusa.learncast.lib.theme.icon.Download
import me.anasmusa.learncast.lib.theme.icon.RemoveDone
import me.anasmusa.learncast.lib.theme.icon.RemoveDownload
import me.anasmusa.learncast.lib.theme.icon.Star
import me.anasmusa.learncast.lib.theme.icon.StarFilled
import me.anasmusa.learncast.string

@Preview
@Composable
private fun LessonPlayerActionSheetPreview() {
    AppTheme {
        PlayerActionSheet(
            isSnip = false,
            downloadState = DownloadState.STOPPED,
            percentDownloaded = 0f,
            isCompleted = false,
            isFavourite = true,
            onDismissRequest = {},
            onDownloadClicked = {},
            onCompletedClicked = {},
            onFavouriteClicked = {},
            onDeleteClicked = {},
        )
    }
}

@Preview
@Composable
private fun SnipPlayerActionSheetPreview() {
    AppTheme {
        PlayerActionSheet(
            isSnip = true,
            downloadState = DownloadState.STOPPED,
            percentDownloaded = 0f,
            isCompleted = false,
            isFavourite = true,
            onDismissRequest = {},
            onDownloadClicked = {},
            onCompletedClicked = {},
            onFavouriteClicked = {},
            onDeleteClicked = {},
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PlayerActionSheet(
    isSnip: Boolean,
    downloadState: DownloadState?,
    percentDownloaded: Float,
    isCompleted: Boolean,
    isFavourite: Boolean,
    onDismissRequest: () -> Unit,
    onDownloadClicked: () -> Unit,
    onCompletedClicked: () -> Unit,
    onFavouriteClicked: () -> Unit,
    onDeleteClicked: () -> Unit,
) {
    ModalBottomSheet(
        modifier =
            Modifier
                .fillMaxWidth(),
        sheetState =
            rememberModalBottomSheetState(
                skipPartiallyExpanded = true,
            ),
        onDismissRequest = onDismissRequest,
    ) {
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(
                        horizontal = 12.dp,
                    ),
        ) {
            Column(
                modifier =
                    Modifier
                        .clip(RoundedCornerShape(8.dp)),
            ) {
                SheetMenuButton(
                    icon =
                        when (downloadState) {
                            null, DownloadState.STOPPED, DownloadState.REMOVING -> Download
                            DownloadState.DOWNLOADING -> Download
                            DownloadState.COMPLETED -> RemoveDownload
                        },
                    title =
                        when (downloadState) {
                            null, DownloadState.REMOVING -> Strings.DOWNLOAD.string()
                            DownloadState.STOPPED -> Strings.RESUME_DOWNLOAD.string()
                            DownloadState.DOWNLOADING -> Strings.DOWNLOADING.string(percentDownloaded.toInt())
                            DownloadState.COMPLETED -> Strings.REMOVE_DOWNLOAD.string()
                        },
                    clip = false,
                    paddingBetween = 28.dp,
                    onClick = {
                        onDismissRequest()
                        onDownloadClicked()
                    },
                )

                if (!isSnip) {
                    SheetMenuButton(
                        icon = if (isCompleted) RemoveDone else DoneAll,
                        title = if (isCompleted) Strings.MARK_NOT_PLAYED.string() else Strings.MARK_COMPLETED.string(),
                        clip = false,
                        paddingBetween = 28.dp,
                        onClick = {
                            onDismissRequest()
                            onCompletedClicked()
                        },
                    )
                    SheetMenuButton(
                        icon = if (isFavourite) StarFilled else Star,
                        title = if (isFavourite) Strings.LESSON_IS_FAVOURITE.string() else Strings.ADD_FAVOURITE.string(),
                        clip = false,
                        paddingBetween = 28.dp,
                        onClick = {
                            onDismissRequest()
                            onFavouriteClicked()
                        },
                    )
                }
            }

            if (isSnip) {
                Spacer(modifier = Modifier.height(12.dp))

                SheetMenuButton(
                    icon = Delete,
                    title = Strings.DELETE.string(),
                    paddingBetween = 28.dp,
                    onClick = {
                        onDismissRequest()
                        onDeleteClicked()
                    },
                )
            }

            Spacer(modifier = Modifier.height(12.dp))
        }
    }
}

File: ios/lib/screen/player/PlayerActionSheet.swift
//
//  PlayerActionSheet.swift
//  ios
//
//  Created by Anas Erkinjonov on 03/02/26.
//

internal import Shared
import SwiftUI

struct PlayerActionSheet: View {
    let isSnip: Bool
    let downloadState: DownloadState?
    let percentDownloaded: Float
    let isCompleted: Bool
    let isFavourite: Bool
    let onDismissRequest: () -> Void
    let onDownloadClicked: () -> Void
    let onRemoveDownloadClicked: () -> Void
    let onCompletedClicked: () -> Void
    let onFavouriteClicked: () -> Void
    let onDeleteClicked: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            // Main actions section
            VStack(spacing: 0) {
                // Download action
                SheetMenuButton(
                    icon: downloadIcon,
                    title: downloadTitle,
                    clip: false,
                    spacing: 28,
                    onClick: {
                        onDismissRequest()
                        handleDownloadAction()
                    }
                )

                // Mark as played/not played (only for non-snips)
                if !isSnip {
                    SheetMenuButton(
                        icon: isCompleted ? "checkmark.circle.badge.xmark" : "checkmark.circle",
                        title: isCompleted ? Strings.shared.MARK_NOT_PLAYED.string() : Strings.shared.MARK_COMPLETED.string(),
                        clip: false,
                        spacing: 28,
                        onClick: {
                            onDismissRequest()
                            onCompletedClicked()
                        }
                    )

                    // Favourite action
                    SheetMenuButton(
                        icon: isFavourite ? "star.fill" : "star",
                        title: isFavourite ? Strings.shared.LESSON_IS_FAVOURITE.string() : Strings.shared.ADD_FAVOURITE.string(),
                        clip: false,
                        spacing: 28,
                        onClick: {
                            onDismissRequest()
                            onFavouriteClicked()
                        }
                    )
                }
            }
            .clipShape(RoundedRectangle(cornerRadius: 12))

            // Delete action (only for snips)
            if isSnip {
                SheetMenuButton(
                    icon: "trash",
                    title: Strings.shared.DELETE.string(),
                    spacing: 28,
                    onClick: {
                        onDismissRequest()
                        onDeleteClicked()
                    }
                )
                .padding(.top, 12)
            }
        }
        .padding(.horizontal, 12)
        .padding(.bottom, 12)
        .padding(.top, 36)
        .presentationBackground(Colors.surfaceContainerLow)
        .presentationDetents([.height(sheetHeight)])
        .presentationDragIndicator(.visible)
    }

    // MARK: - Download State Logic
    private var downloadIcon: String {
        switch downloadState {
        case .none, .stopped, .removing:
            return "arrow.down.circle"
        case .downloading:
            return "arrow.down.circle.fill"
        case .completed:
            return "xmark.circle"
        }
    }

    private var downloadTitle: String {
        switch downloadState {
        case .none, .removing:
            return Strings.shared.DOWNLOAD.string()
        case .stopped:
            return Strings.shared.RESUME_DOWNLOAD.string()
        case .downloading:
            return Strings.shared.DOWNLOADING.string(NSNumber(value: Int(percentDownloaded)))
        case .completed:
            return Strings.shared.REMOVE_DOWNLOAD.string()
        }
    }

    private func handleDownloadAction() {
        switch downloadState {
        case .none, .stopped, .downloading:
            onDownloadClicked()
        case .completed:
            onRemoveDownloadClicked()
        default:
            break  // Do nothing for downloading/removing states
        }
    }

    // MARK: - Sheet Height Calculation
    private var sheetHeight: CGFloat {
        let baseItemHeight: CGFloat = 56
        var itemCount: CGFloat = 1  // Download always present

        if !isSnip {
            itemCount += 2  // Completed + Favourite
        } else {
            itemCount += 1  // Delete button with spacing
        }

        return (itemCount * baseItemHeight) + 40  // 40 for padding
    }
}

// MARK: - Preview
#Preview("Lesson Action Sheet") {
    PreviewRoot {
        Color.black.opacity(0.3)
            .sheet(isPresented: .constant(true)) {
                PlayerActionSheet(
                    isSnip: false,
                    downloadState: .stopped,
                    percentDownloaded: 0,
                    isCompleted: false,
                    isFavourite: true,
                    onDismissRequest: {},
                    onDownloadClicked: {},
                    onRemoveDownloadClicked: {},
                    onCompletedClicked: {},
                    onFavouriteClicked: {},
                    onDeleteClicked: {}
                )
            }
    }
}

#Preview("Snip Action Sheet") {
    PreviewRoot {
        Color.black.opacity(0.3)
            .sheet(isPresented: .constant(true)) {
                PlayerActionSheet(
                    isSnip: true,
                    downloadState: .completed,
                    percentDownloaded: 100,
                    isCompleted: false,
                    isFavourite: false,
                    onDismissRequest: {},
                    onDownloadClicked: {},
                    onRemoveDownloadClicked: {},
                    onCompletedClicked: {},
                    onFavouriteClicked: {},
                    onDeleteClicked: {}
                )
            }
    }
}

#Preview("Downloading State") {
    PreviewRoot {
        Color.black.opacity(0.3)
            .sheet(isPresented: .constant(true)) {
                PlayerActionSheet(
                    isSnip: false,
                    downloadState: .downloading,
                    percentDownloaded: 45,
                    isCompleted: true,
                    isFavourite: false,
                    onDismissRequest: {},
                    onDownloadClicked: {},
                    onRemoveDownloadClicked: {},
                    onCompletedClicked: {},
                    onFavouriteClicked: {},
                    onDeleteClicked: {}
                )
            }
    }
}
